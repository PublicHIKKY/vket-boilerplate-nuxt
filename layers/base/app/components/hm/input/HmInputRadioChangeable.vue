<template>
  <div class="hm-input-radio-changeable">
    <div
      v-for="(option, index) in props.options"
      :key="option.value"
      ref="radiobuttons"
      class="radio"
    >
      <HaBaseInput
        :id="option.value"
        class="input"
        type="radio"
        :name="props.name"
        :value="option.value"
        :model-value="option.value"
        :checked="option.checked"
        :disabled="option.disabled"
        required
        @change="onChange($event)"
      />
      <label
        :for="option.value"
        class="label"
        :class="`option-${index}`"
      >
        {{ option.label }}
      </label>
    </div>
  </div>
</template>

<script lang="ts" setup>
import { z } from 'zod'

type Radio = {
  label: string
  value: string
  checked?: boolean
  disabled?: boolean
}

type Props = {
  name: string
  options: Radio[]
}
const props = defineProps<Props>()

const radiobuttons = ref<HTMLDivElement[]>()

/** props.optionsを監視し、親コンポーネントでの変更をラジオボタンに反映する */
watch(toRef(props.options), (_next, _prev) => {
  // チェックされているオブジェクトを探す
  const checkedOptions
    = props.options.find(element => element.checked)
      ?? raiseError('HmInputRadioChangeable: watch: checkedOptions')

  // チェック対象を探す
  const buttons
    = radiobuttons.value
      ?? raiseError('HmInputRadioChangeable: watch: radiobuttons')
  const checkTarget
    = buttons.find(
      // チェックされているオブジェクトとidが同じものがチェック対象
      element => element.children[0]?.id === checkedOptions?.value,
    ) ?? raiseError('HmInputRadioChangeable: watch: checkTarget')

  // 探したチェック対象をチェック済にする
  const checkbox = z
    .object({ checked: z.boolean() })
    .parse(checkTarget.children[0])
  checkbox.checked = true
})

type Emits = {
  (e: 'change', value: string): void
}
const emit = defineEmits<Emits>()
const onChange = (e: Event) => {
  if (e.target instanceof HTMLInputElement) {
    emit('change', e.target.value)
  }
}
</script>

<style lang="scss" scoped>
@use '#base/app/assets/styles/variables' as v;

.hm-input-radio-changeable {
  display: flex;
  width: 100%;

  .radio {
    flex: 1;

    > .label {
      background-color: v.$navy-1;
      border: solid 1px v.$navy-2;
      cursor: pointer;
      display: block;
      height: 100%;
      padding: v.space(2) 0;
      text-align: center;
      user-select: none;
      white-space: pre-wrap;

      &:hover {
        background-color: v.$green-4;
      }
    }
  }
}

.input {
  display: none;

  &:checked,
  &:hover,
  &:focus {
    + .label {
      background-color: v.$green-4;
      border-color: v.$blue;
    }
  }
}
</style>
