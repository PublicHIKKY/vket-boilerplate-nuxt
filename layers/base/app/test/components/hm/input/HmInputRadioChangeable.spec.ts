import { mount } from '@vue/test-utils'
import HmInputRadioChangeable from '#base/app/components/hm/input/HmInputRadioChangeable.vue'

test('ref component', () => {
  expect(HmInputRadioChangeable).toBeTruthy()
})

test('mount component', () => {
  const wrapper = mount(HmInputRadioChangeable, {
    props: {
      name: 'testName',
      options: [
        {
          label: 'testLabel',
          value: 'testValue',
        },
      ],
    },
  })
  expect(wrapper.getCurrentComponent()).toBeTruthy()
  expect(wrapper.html()).toMatchSnapshot()
})

describe('props', () => {
  it(':name, :value', () => {
    const wrapper = mount(HmInputRadioChangeable, {
      props: {
        name: 'testName',
        options: [
          {
            label: 'testLabel',
            value: 'testValue',
          },
        ],
      },
    })
    expect(wrapper.find('input[type="radio"]').attributes('name')).toBe(
      'testName',
    )
    expect(wrapper.find('label.label').attributes('for')).toBe(
      'testValue',
    )
    expect(wrapper.find('label.label').text()).toBe('testLabel')
  })

  it(':checked', () => {
    const wrapper = mount(HmInputRadioChangeable, {
      props: {
        name: 'testName',
        options: [
          {
            label: 'testLabel',
            value: 'testValue',
            checked: true,
          },
        ],
      },
    })
    expect(
      (wrapper.find('input[type="radio"]').element as HTMLInputElement).checked,
    ).toBeTruthy()
  })

  it(':disabled', () => {
    const wrapper = mount(HmInputRadioChangeable, {
      props: {
        name: 'testName',
        options: [
          {
            label: 'testLabel',
            value: 'testValue',
            checked: true,
            disabled: true,
          },
        ],
      },
    })
    expect(
      (wrapper.find('input[type="radio"]').element as HTMLInputElement).disabled,
    ).toBeTruthy()
  })
})

describe('emits', () => {
  it(':change', async () => {
    const wrapper = mount(HmInputRadioChangeable, {
      props: {
        name: 'testName',
        options: [
          {
            label: 'testLabel',
            value: 'testValue',
            checked: true,
          },
        ],
      },
    })
    await wrapper.find('input[type="radio"]').trigger('change')
    setTimeout(() => {
      expect(wrapper.emitted()).toHaveProperty('change')
      expect(wrapper.emitted()['change']).toHaveLength(1)
      expect(wrapper.emitted()['change']).toEqual([['testValue']])
    }, 1)
  })
})

// NOTE: 「props.optionsを監視し、親コンポーネントでの変更をラジオボタンに反映する」というコンポーネント内のwatchの動作をテスト
describe('DOM check', () => {
  it(':Change radio button based on parent props', async () => {
    const wrapper = mount(HmInputRadioChangeable, {
      props: {
        name: 'testName',
        options: [
          {
            label: 'testLabel1',
            value: 'testValue1',
            checked: true,
          },
          {
            label: 'testLabel2',
            value: 'testValue2',
            checked: false,
          },
          {
            label: 'testLabel3',
            value: 'testValue3',
            checked: false,
          },
        ],
      },
    })
    expect(
      (wrapper.find('input[id="testValue1"]').element as HTMLInputElement)
        .checked,
    ).toBeTruthy()
    expect(
      (wrapper.find('input[id="testValue2"]').element as HTMLInputElement)
        .checked,
    ).toBeFalsy()
    expect(
      (wrapper.find('input[id="testValue3"]').element as HTMLInputElement)
        .checked,
    ).toBeFalsy()
    // NOTE: props,optionsを変えてcheckedを再確認
    await wrapper.setProps({
      options: [
        {
          label: 'testLabel1',
          value: 'testValue1',
          checked: false,
        },
        {
          label: 'testLabel2',
          value: 'testValue2',
          checked: true,
        },
        {
          label: 'testLabel3',
          value: 'testValue3',
          checked: false,
        },
      ],
    })
    expect(
      (wrapper.find('input[id="testValue1"]').element as HTMLInputElement)
        .checked,
    ).toBeFalsy()
    expect(
      (wrapper.find('input[id="testValue2"]').element as HTMLInputElement)
        .checked,
    ).toBeTruthy()
    expect(
      (wrapper.find('input[id="testValue3"]').element as HTMLInputElement)
        .checked,
    ).toBeFalsy()
  })
})
