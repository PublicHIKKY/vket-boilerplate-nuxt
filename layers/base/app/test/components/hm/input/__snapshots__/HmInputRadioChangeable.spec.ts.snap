// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`mount component 1`] = `
"<div data-v-0195f537="" class="hm-input-radio-changeable">
  <div data-v-0195f537="" class="radio"><input data-v-d0a26e0a="" data-v-0195f537="" id="testValue" class="ha-base-input input" type="radio" name="testName" required="" value="testValue"><label data-v-0195f537="" for="testValue" class="label option-0">testLabel</label></div>
</div>"
`;
